---
output:
  pdf_document: default
  html_document: default
---
```{r, echo = FALSE, include = FALSE}
library(data.table)
library(dplyr)
```

I'm trying to understand how the GCTA trait simulation works and am walking through the calculation for a simulated trait. I'm using the example dataset from a simulation run with 96 individuals.

Here is the GWAS simulation notes from GCTA:

"The phenotypes are simulated based on a set of real genotype data and a simple additive genetic model yj = sum(wijui) + ej, where wij = (xij - 2pi) / sqrt[2pi(1 - pi)] with xij being the number of reference alleles for the i-th causal variant of the j-th individual and pi being the frequency of the i-th causal variant, ui is the allelic effect of the i-th causal variant and ej is the residual effect generated from a normal distribution with mean of 0 and variance of var(sum(wij*ui))(1 / h2 - 1)."

I have an example dataset from a simulation run with 96 individuals. A simulated trait (`data/test_data/1_1_gamma_ce.96.allout15_irrepressible.grosbeak_0.05/0.9/GCTA_SIMULATE_PHENOTYPES/1_1_0.9_0.05_gamma_ce.96.allout15_irrepressible.grosbeak_sims.phen`) was generated by the simulation pipeline with a heritability of 0.9.
```{r}
# set path to the genotype matrix created for the test mapping panel
gt_m_path <- "data/test_data/ce.96.allout15_irrepressible.grosbeak_0.05_Genotype_Matrix.tsv"

# path to the file containing the causal variants
causal_var_path <- "data/test_data/1_1_gamma_ce.96.allout15_irrepressible.grosbeak_0.05/PYTHON_SIMULATE_EFFECTS_GLOBAL/causal.variants.sim.1.1.txt"

# path to the .par file for the simulation
par_path <- "data/test_data/1_1_gamma_ce.96.allout15_irrepressible.grosbeak_0.05/0.9/GCTA_SIMULATE_PHENOTYPES/1_1_0.9_0.05_gamma_ce.96.allout15_irrepressible.grosbeak_sims.par"

# path to the simulated phenotype file
phen_path <- "data/test_data/1_1_gamma_ce.96.allout15_irrepressible.grosbeak_0.05/0.9/GCTA_SIMULATE_PHENOTYPES/1_1_0.9_0.05_gamma_ce.96.allout15_irrepressible.grosbeak_sims.phen"
```

```{r}
# load the causal variants
causal_var <- data.table::fread(
  causal_var_path,
  col.names = c("snpid", "effect")
) %>% 
  tidyr::separate(snpid, into = c("chrom", "pos"), sep = ":", convert = TRUE)
head(causal_var)
```

The causal variant selected is 1:4644600 and has a frequency of 0.0833333 and the effect size assigned to it from the gamma distribution is 2.48992.
This is included in the par file generated by the GCTA trait simulation step of the pipeline.
```{r}
# load the par file
par <- data.table::fread(par_path, header = TRUE, sep = "\t", data.table = FALSE)
head(par)
```
Load the genotype matrix
```{r, include = FALSE}
# load the genotype matrix
gt_m <- data.table::fread(gt_m_path, header = TRUE, sep = "\t", data.table = FALSE)
```
# Get the genotype matrix for the causal variant
```{r}
cv_genos <- gt_m %>%
  dplyr::filter(
    CHROM == causal_var$chrom[1],
    POS == causal_var$pos[1]
  )
head(cv_genos)
```
Drop the `CHOM`, `POS`, `REF` and `ALT` columns of the genotype matrix and create a new data frame by pivoting the genotype matrix long
```{r}
# drop the CHROM, POS, REF and ALT columns
cv_genos <- cv_genos %>%
  dplyr::select(-CHROM, -POS, -REF, -ALT) %>%
  tidyr::pivot_longer(
    cols = everything(),
    names_to = "strain",
    values_to = "GT"
  )
head(cv_genos)
```

Based on the `GT` column we are going to create a new colum that shows the number of reference alleles for each strain. If a strain is -1 (homozygous REF) then the number of reference alleles is 0. If a strain is 1 (homozygous ALT) then the number of reference alleles is 2. 
```{r}
cv_genos <- cv_genos %>%
  dplyr::mutate(
    REF_ALLELES = dplyr::case_when(
      GT == -1 ~ 0,
      GT == 1 ~ 2,
      TRUE ~ NA_real_
    )
  )
head(cv_genos)
```

# Calculate the wij for all strains based on the number of reference alleles and the AF
Pull / set the variables for the calculation
```{r}
# get the AF for the SNP of interest
p <- par %>%
  dplyr::pull(Frequency)
# get the effect size for the SNP of interest
u <- effect_size <- causal_var %>%
  dplyr::pull(effect)

h2 <- 0.9

print(glue::glue("AF: {p}"))
print(glue::glue("Effect size: {u}"))
print(glue::glue("Heritability: {h2}"))
```
Now we can calculate the wij for all strains based on the number of reference alleles and the AF using the formula wij = (REF_ALLELES - 2 * p) / sqrt(2 * p * (1 - p)) 
```{r}
# calculate the wij for all strains based on the number of reference alleles and the AF
cv_genos_wij <- cv_genos %>%
  dplyr::mutate(
    wij = (REF_ALLELES - 2 * p) / sqrt(2 * p * (1 - p))
  )
head(cv_genos_wij)
```
Now we can calculate the genetic component for all strains based on the wij and the effect size using the formula gj = wij * u
```{r}
# calculate the genetic component for all strains based on the wij and the effect size
cv_genos_gj <- cv_genos_wij %>%
  dplyr::mutate(
    gj = wij * u
  )
head(cv_genos_gj)
```
Look at the genetic component for strains with the causal variant
```{r}
# look at the genetic component for strains with the causal variant
cv_genos_gj %>%
  dplyr::filter(REF_ALLELES == 2) %>% 
  head()
```
# Calculate the residual effect
To get the residual effect we need to calculate the variance of the genetic component using the formula var(gj) = var(wij * u) 
```{r}
# calculate the variance of the genetic component
var_g <- var(cv_genos_gj$gj)
print(var_g)
```
Now we can calculate the residual effect using the formula var(e) = var(g) * (1 / h2 - 1)
```{r}
# calculate the residual effect
var_e <- var_g * (1 / h2 - 1)
print(head(var_e))
```
Now we can sample from a normal distribution to get the residual effect using the formula ej = rnorm(n = 96, mean = 0, sd = sqrt(var_e))
```{r}
# sample from a normal distribution to get the residual effect
set.seed(123)
ej <- rnorm(n = 96, mean = 0, sd = sqrt(var_e))
print(ej)
```
Lets also look at the range of the residual effect
```{r}
# look at the range of the residual effect
residual_effect_range  <- range(ej)
print(residual_effect_range)
```
# Calculate the trait value for all strains
Now we can calculate the phenotype for all strains using the formula phen = gj + ej
```{r}
# calculate the phenotype for all strains
cv_genos_phen <- cv_genos_gj %>%
  dplyr::mutate(
    phen = gj + ej
  )
head(cv_genos_phen)
```
Peak at the phenotype for strains with the causal variant
```{r}
# peak at the phenotype for strains with the causal variant
cv_genos_phen %>%
  dplyr::filter(REF_ALLELES == 2) %>% 
  head()
```
# Compare the calculated phenotype to the simulated phenotype
And compare it to the simulated phenotype file
```{r}
# compare it to the simulated phenotype file
phen <- data.table::fread(
  phen_path,
  col.names = c("id", "strain", "simulated_pheno")
) %>% 
  dplyr::select(-id)
head(phen)
```
Attach the phenotype to the genotype matrix
```{r}
# attach the phenotype to the genotype matrix
cv_genos_phen_check <- cv_genos_phen %>%
  dplyr::left_join(phen, by = "strain")
head(cv_genos_phen_check)
```

Check the strains with only alt alleles
```{r}
# check the strains with no reference alleles
cv_genos_phen_check %>%
  dplyr::filter(REF_ALLELES == 2)
```
Lets make one final check that the difference between the simulated phenotype and the calculated phenotype is due to the residual effect
The abs difference should be within the residual effect range since the residual effects are randomly assigned to each strain
```{r}
cv_genos_phen_check <- cv_genos_phen_check %>%
  dplyr::mutate(
    abs_diff = abs(simulated_pheno - phen)
  ) %>% 
  arrange(desc(abs_diff))

head(cv_genos_phen_check)
```
Calculate the mean of the absolute difference between the simulated phenotype and the calculated phenotype
```{r}
# calcualte the mean of the absolute difference between the simulated phenotype and the calculated phenotype
mean_abs_diff <- mean(cv_genos_phen_check$abs_diff)
print(glue::glue("Mean absolute difference: {mean_abs_diff}"))

range_abs_diff <- range(cv_genos_phen_check$abs_diff)
print(glue::glue("Range of absolute difference: {range_abs_diff}")) 
```


